In [51]:
import pandas as pd
import numpy as np

In [52]:
dataset = pd.read_csv('BostonHousing.csv')
dataset

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     ptratio       b  lstat  medv  
0       15.3  396.90   4.98  24.0  
1       17.8  396.90   9.14  21.6  
2       17.8  392.83   4.03  34.7  
3       18.7  394.63   2.94  33.4  
4       18.7  396.90   5.33  36.2  
..       ...     ...    ...   ...  
501     21.0  391.99   9.67  22.4  
502     21.0  396.90   9.08  20.6  
503     21.0  396.90   5.64  23.9  
504     21.0  393.45   6.48  22.0  
505     21.0  396.90   7.88  11.9  

[506 rows x 14 columns]

In [53]:
x = dataset.iloc[:,:-1]
x

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  \
0    0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296   
1    0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242   
2    0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242   
3    0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222   
4    0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222   
..       ...   ...    ...   ...    ...    ...   ...     ...  ...  ...   
501  0.06263   0.0  11.93     0  0.573  6.593  69.1  2.4786    1  273   
502  0.04527   0.0  11.93     0  0.573  6.120  76.7  2.2875    1  273   
503  0.06076   0.0  11.93     0  0.573  6.976  91.0  2.1675    1  273   
504  0.10959   0.0  11.93     0  0.573  6.794  89.3  2.3889    1  273   
505  0.04741   0.0  11.93     0  0.573  6.030  80.8  2.5050    1  273   

     ptratio       b  lstat  
0       15.3  396.90   4.98  
1       17.8  396.90   9.14  
2       17.8  392.83   4.03  
3       18.7  394.63   2.94  
4       18.7  396.90   5.33  
..       ...     ...    ...  
501     21.0  391.99   9.67  
502     21.0  396.90   9.08  
503     21.0  396.90   5.64  
504     21.0  393.45   6.48  
505     21.0  396.90   7.88  

[506 rows x 13 columns]

In [54]:
y = dataset.iloc[:,-1]
y

0      24.0
1      21.6
2      34.7
3      33.4
4      36.2
       ... 
501    22.4
502    20.6
503    23.9
504    22.0
505    11.9
Name: medv, Length: 506, dtype: float64

In [55]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=5)

In [56]:
x_train

crim    zn  indus  chas    nox     rm    age      dis  rad  tax  \
33    1.15172   0.0   8.14     0  0.538  5.701   95.0   3.7872    4  307   
283   0.01501  90.0   1.21     1  0.401  7.923   24.8   5.8850    1  198   
418  73.53410   0.0  18.10     0  0.679  5.957  100.0   1.8026   24  666   
502   0.04527   0.0  11.93     0  0.573  6.120   76.7   2.2875    1  273   
402   9.59571   0.0  18.10     0  0.693  6.404  100.0   1.6390   24  666   
..        ...   ...    ...   ...    ...    ...    ...      ...  ...  ...   
486   5.69175   0.0  18.10     0  0.583  6.114   79.8   3.5459   24  666   
189   0.08370  45.0   3.44     0  0.437  7.185   38.9   4.5667    5  398   
495   0.17899   0.0   9.69     0  0.585  5.670   28.8   2.7986    6  391   
206   0.22969   0.0  10.59     0  0.489  6.326   52.5   4.3549    4  277   
355   0.10659  80.0   1.91     0  0.413  5.936   19.5  10.5857    4  334   

     ptratio       b  lstat  
33      21.0  358.77  18.35  
283     13.6  395.52   3.16  
418     20.2   16.45  20.62  
502     21.0  396.90   9.08  
402     20.2  376.11  20.31  
..       ...     ...    ...  
486     20.2  392.68  14.98  
189     15.2  396.90   5.39  
495     19.2  393.29  17.60  
206     18.6  394.87  10.97  
355     22.0  376.04   5.57  

[404 rows x 13 columns]

In [57]:
y_train

33     13.1
283    50.0
418     8.8
502    20.6
402    12.1
       ... 
486    19.1
189    34.9
495    23.1
206    24.4
355    20.6
Name: medv, Length: 404, dtype: float64

In [58]:
x_test

crim    zn  indus  chas    nox     rm    age     dis  rad  tax  \
226   0.38214   0.0   6.20     0  0.504  8.040   86.5  3.2157    8  307   
292   0.03615  80.0   4.95     0  0.411  6.630   23.4  5.1167    4  245   
90    0.04684   0.0   3.41     0  0.489  6.417   66.1  3.0923    2  270   
373  11.10810   0.0  18.10     0  0.668  4.906  100.0  1.1742   24  666   
273   0.22188  20.0   6.96     1  0.464  7.691   51.8  4.3665    3  223   
..        ...   ...    ...   ...    ...    ...    ...     ...  ...  ...   
349   0.02899  40.0   1.25     0  0.429  6.939   34.5  8.7921    1  335   
212   0.21719   0.0  10.59     1  0.489  5.807   53.8  3.6526    4  277   
156   2.44668   0.0  19.58     0  0.871  5.272   94.0  1.7364    5  403   
480   5.82401   0.0  18.10     0  0.532  6.242   64.7  3.4242   24  666   
248   0.16439  22.0   5.86     0  0.431  6.433   49.1  7.8265    7  330   

     ptratio       b  lstat  
226     17.4  387.38   3.13  
292     19.2  396.90   4.70  
90      17.8  392.18   8.81  
373     20.2  396.90  34.77  
273     18.6  390.77   6.58  
..       ...     ...    ...  
349     19.7  389.85   5.89  
212     18.6  390.94  16.03  
156     14.7   88.63  16.14  
480     20.2  396.90  10.74  
248     19.1  374.71   9.52  

[102 rows x 13 columns]

In [59]:
y_test

226    37.6
292    27.9
90     22.6
373    13.8
273    35.2
       ... 
349    26.6
212    22.4
156    13.1
480    23.0
248    24.5
Name: medv, Length: 102, dtype: float64

In [60]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()

In [61]:
x_train = sc.fit_transform(x_train)
x_train

array([[-2.76469903e-01, -5.07766334e-01, -4.02011945e-01, ...,
         1.17561917e+00, -3.01968083e-04,  8.35287872e-01],
       [-4.07463681e-01,  3.33136528e+00, -1.43188822e+00, ...,
        -2.20285923e+00,  4.12907260e-01, -1.32549534e+00],
       [ 8.06483214e+00, -5.07766334e-01,  1.07815655e+00, ...,
         8.10378266e-01, -3.84927565e+00,  1.15819623e+00],
       ...,
       [-3.88566725e-01, -5.07766334e-01, -1.71664438e-01, ...,
         3.53827130e-01,  3.87833612e-01,  7.28600090e-01],
       [-3.82724087e-01, -5.07766334e-01, -3.79142727e-02, ...,
         7.98964487e-02,  4.05598798e-01, -2.14519905e-01],
       [-3.96910057e-01,  2.90479510e+00, -1.32786031e+00, ...,
         1.63217031e+00,  1.93878260e-01, -9.82671938e-01]])

In [62]:
x_test = sc.transform(x_test)
x_test

array([[-0.36515584, -0.50776633, -0.69031786, ..., -0.46796491,
         0.32138282, -1.32976286],
       [-0.40502752,  2.9047951 , -0.87608198, ...,  0.35382713,
         0.42842369, -1.10642977],
       [-0.40379561, -0.50776633, -1.10494337, ..., -0.28534446,
         0.37535301, -0.52178072],
       ...,
       [-0.12723949, -0.50776633,  1.29810127, ..., -1.70065298,
        -3.037699  ,  0.52091454],
       [ 0.26196199, -0.50776633,  1.07815655, ...,  0.81037827,
         0.42842369, -0.24723749],
       [-0.39024922,  0.43068806, -0.7408457 , ...,  0.30817202,
         0.17892402, -0.42078295]])

In [63]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(x_train,y_train)
y_pred = model.predict(x_test)
y_pred

array([37.56311787, 32.14445143, 27.06573629,  5.67080633, 35.09982577,
        5.85803701, 27.53708506, 31.81019188, 26.35634771, 22.77208748,
       31.91183048, 21.50224061, 23.70119983, 33.3622504 , 28.51633591,
       14.39456899,  0.19284025, 18.66247155, 13.71004139, 14.13408635,
        2.03263952, 19.7280831 , 38.18657429, 24.19760058, 31.30247973,
       11.14144544, 25.03636951, 23.27970871, 22.49420127, 20.52972594,
       15.16513744,  6.92553586, 18.3557733 , 22.37179804, 28.91287973,
       19.02980786, 30.19357214,  8.74384915, 40.86691522, 34.53763591,
       20.70224878,  2.59618963, 29.99590282, 12.15704798, 27.10186397,
       30.8052437 , -6.24169079, 19.84885777, 20.92973441, 12.43523958,
       20.4949947 , 19.19231742, 23.69073157, 12.67998473, 17.14252424,
       25.04649176, 34.77758126, 15.23294903, 28.22306193, 21.08745388,
       20.39506129, 25.79476888, 14.72463673, 33.18635032, 23.17771307,
       13.11057248, 19.23154617, 24.61162961, 21.50327036, 22.00

In [64]:
from sklearn.model_selection import cross_val_score
print("Training:", np.mean(cross_val_score(model,x_train,y_train,cv=10,scoring='r2')))
print("Testing:", np.mean(cross_val_score(model,x_test,y_test,cv=10,scoring='r2')))

Training: 0.6965014634091666
Testing: 0.7016177881094494


#### We can clearly see that there is a large gap in explainablity of variance in training and testing data. so, we can say that there is an overfitting in model.

In [65]:
from mlxtend.feature_selection import SequentialFeatureSelector as SFS
sfs = SFS(model,k_features='best',forward = False,floating = False, scoring = 'r2',cv = 10)

In [66]:
sel = sfs.fit(x_train,y_train)

In [67]:
sel.k_feature_names_

('0', '1', '4', '5', '7', '8', '9', '10', '11', '12')

In [68]:
x_sel = sel.subsets_
x_sel = pd.DataFrame(x_sel).T
x_sel

feature_idx  \
13  (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12)   
12     (0, 1, 2, 3, 4, 5, 7, 8, 9, 10, 11, 12)   
11        (0, 1, 2, 4, 5, 7, 8, 9, 10, 11, 12)   
10           (0, 1, 4, 5, 7, 8, 9, 10, 11, 12)   
9               (0, 4, 5, 7, 8, 9, 10, 11, 12)   
8                  (0, 4, 5, 7, 8, 10, 11, 12)   
7                     (4, 5, 7, 8, 10, 11, 12)   
6                        (4, 5, 7, 10, 11, 12)   
5                           (5, 7, 10, 11, 12)   
4                               (5, 7, 10, 12)   
3                                  (5, 10, 12)   
2                                      (5, 12)   
1                                        (12,)   

                                            cv_scores avg_score  \
13  [0.7042310425171581, 0.6078393010267444, 0.734...  0.696501   
12  [0.7058578946614813, 0.6111956117116959, 0.734...  0.700082   
11  [0.6977135685685397, 0.5893451722940015, 0.756...  0.703106   
10  [0.6976271527384033, 0.5893677473731809, 0.756...  0.704271   
9   [0.6846552899687137, 0.5676582720229011, 0.747...  0.699583   
8   [0.6848824072673403, 0.557717062579016, 0.7332...  0.695145   
7   [0.6970093900586629, 0.551601766684709, 0.7046...  0.686324   
6   [0.6844477223042464, 0.5269907647025986, 0.717...  0.683571   
5   [0.6735730731603646, 0.5528003905977739, 0.717...   0.67722   
4   [0.6691142989449639, 0.5283756637685816, 0.726...  0.666096   
3   [0.6349864106952552, 0.48860915105956604, 0.66...  0.654299   
2   [0.6094937603864676, 0.5193055192284901, 0.653...  0.613085   
1   [0.579600734688748, 0.5643891909094458, 0.6106...  0.530034   

                                 feature_names  
13  (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12)  
12     (0, 1, 2, 3, 4, 5, 7, 8, 9, 10, 11, 12)  
11        (0, 1, 2, 4, 5, 7, 8, 9, 10, 11, 12)  
10           (0, 1, 4, 5, 7, 8, 9, 10, 11, 12)  
9               (0, 4, 5, 7, 8, 9, 10, 11, 12)  
8                  (0, 4, 5, 7, 8, 10, 11, 12)  
7                     (4, 5, 7, 8, 10, 11, 12)  
6                        (4, 5, 7, 10, 11, 12)  
5                           (5, 7, 10, 11, 12)  
4                               (5, 7, 10, 12)  
3                                  (5, 10, 12)  
2                                      (5, 12)  
1                                        (12,)

In [69]:
def Adjusted_R2(r2,n,k):
    return 1- ((1-r2)*(n-1)/(n-k-1))

In [70]:
x_sel['total rows'] = x_train.shape[0]
x_sel['no. of features'] = x_sel['feature_idx'].apply(lambda x:len(x))
x_sel['Adjusted R2'] = Adjusted_R2(x_sel['avg_score'],x_sel['total rows'],x_sel['no. of features'])
x_sel.sort_values('Adjusted R2',ascending=False)

feature_idx  \
10           (0, 1, 4, 5, 7, 8, 9, 10, 11, 12)   
11        (0, 1, 2, 4, 5, 7, 8, 9, 10, 11, 12)   
9               (0, 4, 5, 7, 8, 9, 10, 11, 12)   
12     (0, 1, 2, 3, 4, 5, 7, 8, 9, 10, 11, 12)   
8                  (0, 4, 5, 7, 8, 10, 11, 12)   
13  (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12)   
7                     (4, 5, 7, 8, 10, 11, 12)   
6                        (4, 5, 7, 10, 11, 12)   
5                           (5, 7, 10, 11, 12)   
4                               (5, 7, 10, 12)   
3                                  (5, 10, 12)   
2                                      (5, 12)   
1                                        (12,)   

                                            cv_scores avg_score  \
10  [0.6976271527384033, 0.5893677473731809, 0.756...  0.704271   
11  [0.6977135685685397, 0.5893451722940015, 0.756...  0.703106   
9   [0.6846552899687137, 0.5676582720229011, 0.747...  0.699583   
12  [0.7058578946614813, 0.6111956117116959, 0.734...  0.700082   
8   [0.6848824072673403, 0.557717062579016, 0.7332...  0.695145   
13  [0.7042310425171581, 0.6078393010267444, 0.734...  0.696501   
7   [0.6970093900586629, 0.551601766684709, 0.7046...  0.686324   
6   [0.6844477223042464, 0.5269907647025986, 0.717...  0.683571   
5   [0.6735730731603646, 0.5528003905977739, 0.717...   0.67722   
4   [0.6691142989449639, 0.5283756637685816, 0.726...  0.666096   
3   [0.6349864106952552, 0.48860915105956604, 0.66...  0.654299   
2   [0.6094937603864676, 0.5193055192284901, 0.653...  0.613085   
1   [0.579600734688748, 0.5643891909094458, 0.6106...  0.530034   

                                 feature_names  total rows  no. of features  \
10           (0, 1, 4, 5, 7, 8, 9, 10, 11, 12)         404               10   
11        (0, 1, 2, 4, 5, 7, 8, 9, 10, 11, 12)         404               11   
9               (0, 4, 5, 7, 8, 9, 10, 11, 12)         404                9   
12     (0, 1, 2, 3, 4, 5, 7, 8, 9, 10, 11, 12)         404               12   
8                  (0, 4, 5, 7, 8, 10, 11, 12)         404                8   
13  (0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12)         404               13   
7                     (4, 5, 7, 8, 10, 11, 12)         404                7   
6                        (4, 5, 7, 10, 11, 12)         404                6   
5                           (5, 7, 10, 11, 12)         404                5   
4                               (5, 7, 10, 12)         404                4   
3                                  (5, 10, 12)         404                3   
2                                      (5, 12)         404                2   
1                                        (12,)         404                1   

   Adjusted R2  
10    0.696746  
11    0.694775  
9     0.692721  
12    0.690878  
8     0.688971  
13    0.686385  
7     0.680779  
6     0.678789  
5     0.673165  
4     0.662749  
3     0.651707  
2     0.611155  
1     0.528865

In [ ]:
x_train = sel.transform(x_train)

array([[-2.76469903e-01, -5.07766334e-01, -1.18484091e-01, ...,
         1.17561917e+00, -3.01968083e-04,  8.35287872e-01],
       [-4.07463681e-01,  3.33136528e+00, -1.31414794e+00, ...,
        -2.20285923e+00,  4.12907260e-01, -1.32549534e+00],
       [ 8.06483214e+00, -5.07766334e-01,  1.11208965e+00, ...,
         8.10378266e-01, -3.84927565e+00,  1.15819623e+00],
       ...,
       [-3.88566725e-01, -5.07766334e-01,  2.91707154e-01, ...,
         3.53827130e-01,  3.87833612e-01,  7.28600090e-01],
       [-3.82724087e-01, -5.07766334e-01, -5.46130284e-01, ...,
         7.98964487e-02,  4.05598798e-01, -2.14519905e-01],
       [-3.96910057e-01,  2.90479510e+00, -1.20941826e+00, ...,
         1.63217031e+00,  1.93878260e-01, -9.82671938e-01]])

In [73]:
x_train = pd.DataFrame(x_train)
x_train

0         1         2         3         4         5         6  \
0   -0.276470 -0.507766 -0.118484 -0.848834 -0.022455 -0.636453 -0.583618   
1   -0.407464  3.331365 -1.314148  2.361809  0.999022 -0.981745 -1.231926   
2    8.064832 -0.507766  1.112090 -0.478931 -0.988813  1.665490  1.551633   
3   -0.403977 -0.507766  0.186977 -0.243407 -0.752702 -0.981745 -0.785842   
4    0.696610 -0.507766  1.234274  0.166954 -1.068474  1.665490  1.551633   
..        ...       ...       ...       ...       ...       ...       ...   
399  0.246720 -0.507766  0.274252 -0.252077 -0.139951  1.665490  1.551633   
400 -0.399548  1.411799 -0.999959  1.295448  0.357105 -0.521356 -0.042371   
401 -0.388567 -0.507766  0.291707 -0.893627 -0.503833 -0.406259 -0.084005   
402 -0.382724 -0.507766 -0.546130  0.054249  0.253974 -0.636453 -0.762051   
403 -0.396910  2.904795 -1.209418 -0.509275  3.287925 -0.636453 -0.423028   

            7         8         9  
0    1.175619 -0.000302  0.835288  
1   -2.202859  0.412907 -1.325495  
2    0.810378 -3.849276  1.158196  
3    1.175619  0.428424 -0.483373  
4    0.810378  0.194665  1.114099  
..        ...       ...       ...  
399  0.810378  0.380975  0.355904  
400 -1.472377  0.428424 -1.008277  
401  0.353827  0.387834  0.728600  
402  0.079896  0.405599 -0.214520  
403  1.632170  0.193878 -0.982672  

[404 rows x 10 columns]

In [75]:
x_test = sel.transform(x_test)
x_test = pd.DataFrame(x_test)
x_test

0         1         2         3         4         5         6  \
0   -0.365156 -0.507766 -0.415218  2.530867 -0.300735 -0.176064 -0.583618   
1   -0.405028  2.904795 -1.226873  0.493509  0.624915 -0.636453 -0.952380   
2   -0.403796 -0.507766 -0.546130  0.185738 -0.360822 -0.866648 -0.803686   
3    0.870897 -0.507766  1.016087 -1.997557 -1.294799  1.665490  1.551633   
4   -0.383624  0.345374 -0.764317  2.026585  0.259622 -0.751550 -1.083231   
..        ...       ...       ...       ...       ...       ...       ...   
97  -0.405853  1.198514 -1.069779  0.939994  2.414571 -0.981745 -0.417080   
98  -0.384165 -0.507766 -0.546130 -0.695671 -0.087996 -0.636453 -0.762051   
99  -0.127239 -0.507766  2.787765 -1.468711 -1.021048 -0.521356 -0.012632   
100  0.261962 -0.507766 -0.170849 -0.067125 -0.199210  1.665490  1.551633   
101 -0.390249  0.430688 -1.052324  0.208857  1.944393 -0.291162 -0.446819   

            7         8         9  
0   -0.467965  0.321383 -1.329763  
1    0.353827  0.428424 -1.106430  
2   -0.285344  0.375353 -0.521781  
3    0.810378  0.428424  3.171039  
4    0.079896  0.359499 -0.838999  
..        ...       ...       ...  
97   0.582103  0.349155 -0.937152  
98   0.079896  0.361411  0.505267  
99  -1.700653 -3.037699  0.520915  
100  0.810378  0.428424 -0.247237  
101  0.308172  0.178924 -0.420783  

[102 rows x 10 columns]

In [77]:
model.fit(x_train,y_train)

LinearRegression()

In [78]:
y_pred_2 = model.predict(x_test)
y_pred_2

array([ 3.79056954e+01,  3.23388433e+01,  2.72933453e+01,  5.75540238e+00,
        3.26284759e+01,  5.84449027e+00,  2.78290569e+01,  2.93852591e+01,
        2.65659587e+01,  2.29170658e+01,  3.20854970e+01,  2.17208360e+01,
        2.39573838e+01,  3.37039191e+01,  2.87386766e+01,  1.42180058e+01,
       -3.05894702e-02,  1.94165287e+01,  1.38669216e+01,  1.39545898e+01,
        1.77331087e+00,  1.98966027e+01,  3.85452081e+01,  2.44475362e+01,
        3.14138291e+01,  1.13524411e+01,  2.52096222e+01,  2.34727402e+01,
        2.26896033e+01,  2.03190555e+01,  1.53992980e+01,  6.73310603e+00,
        1.86226625e+01,  2.26731023e+01,  2.90935893e+01,  1.91910471e+01,
        2.77325352e+01,  8.75121793e+00,  4.15874603e+01,  3.47869325e+01,
        2.10118981e+01,  2.57444046e+00,  3.02917085e+01,  1.23100170e+01,
        2.72073884e+01,  3.08581684e+01, -6.47344937e+00,  2.01524449e+01,
        2.16984771e+01,  1.25230964e+01,  2.05820084e+01,  1.92843442e+01,
        2.36842659e+01,  

In [79]:
print("Training:", np.mean(cross_val_score(model,x_train,y_train,cv=10,scoring='r2')))
print("Testing:", np.mean(cross_val_score(model,x_test,y_test,cv=10,scoring='r2')))

Training: 0.7042709681532112
Testing: 0.7167668610246819


#### Before feature selection 
#### Training Accuracy: 0.6965014634091666 ; Testing Accuracy: 0.7016177881094494
#### After Feature Selection  (SEQUENTIAL BACKWARD SELECTION)
#### Training: 0.7042709681532112 ; Testing: 0.7167668610246819